discovery/interceptor.go

```diff
@@ -0,0 +1,99 @@
+package discovery
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+const tokenField = "x-consul-token"
+
+func makeUnaryInterceptor(watcher *Watcher) grpc.UnaryClientInterceptor {
+	return func(
+		ctx context.Context,
+		method string,
+		req, reply interface{},
+		cc *grpc.ClientConn,
+		invoker grpc.UnaryInvoker,
+		opts ...grpc.CallOption,
+	) error {
+		ctx = interceptContext(watcher, ctx)
+		err := invoker(ctx, method, req, reply, cc, opts...)
+		interceptError(watcher, err)
+		return err
+	}
+}
+
+func makeStreamInterceptor(watcher *Watcher) grpc.StreamClientInterceptor {
+	return func(
+		ctx context.Context,
+		desc *grpc.StreamDesc,
+		cc *grpc.ClientConn,
+		method string,
+		streamer grpc.Streamer,
+		opts ...grpc.CallOption,
+	) (grpc.ClientStream, error) {
+		ctx = interceptContext(watcher, ctx)
+		stream, err := streamer(ctx, desc, cc, method, opts...)
+		interceptError(watcher, err)
+		if err != nil {
+			return nil, err
+		}
+		return &streamWrapper{ClientStream: stream, watcher: watcher}, nil
+	}
+}
+
+type streamWrapper struct {
+	grpc.ClientStream
+	watcher *Watcher
+}
+
+func (s *streamWrapper) SendMsg(m interface{}) error {
+	err := s.ClientStream.SendMsg(m)
+	interceptError(s.watcher, err)
+	return err
+}
+
+func (s *streamWrapper) RecvMsg(m interface{}) error {
+	err := s.ClientStream.RecvMsg(m)
+	interceptError(s.watcher, err)
+	return err
+}
```

Comment:
This is really neat! Do you think we'll retry here after switching servers, or leave that to the caller?

I think it might need to be the latter for stateful streams like xDS's, where I'm wondering if we'll event need to force Envoy to reconnect to reset its nonces.

Reply:
Yep, seems reasonable to me 👍🏻